QEfficient/cloud/infer.py

```diff
@@ -103,61 +85,55 @@ def main(
+
+        # Compile
+        # We need to pass parent directory of qpc_dir_path, as the compile function handles the qpcs directory creation
+        generated_qpc_path = QEfficient.compile(
+            onnx_path=onnx_model_path,
+            qpc_path=os.path.dirname(qpc_dir_path),
+            num_cores=num_cores,
+            batch_size=batch_size,
+            prompt_len=prompt_len,
+            ctx_len=ctx_len,
+            mxfp6=mxfp6,
+            mxint8=mxint8,
+            aic_enable_depth_first=aic_enable_depth_first,
+            mos=mos,
+            device_group=device_group,
         )
-        return
-
-    #############################################
-    # hf model -> export -> compile -> execute
-    #############################################
-    model_hf = AutoModelForCausalLM.from_pretrained(model_hf_path, use_cache=True)
-    # Easy and minimal api to update the model to QEff.
-    model_transformed = QEfficient.transform(model_hf, type="Transformers", form_factor="cloud")
-    logger.info(f"Model after Optimized transformations {model_transformed}")
-
-    # Export to the Onnx
-    logger.info(f"Exporting to Pytorch {model_name} to ONNX...")
-    base_path, generated_onnx_path = qualcomm_efficient_converter(
-        model_kv=model_transformed,
-        onnx_dir_path=onnx_dir_path,
-        model_name=model_name,
-        kv=True,
-        form_factor="cloud",
-        return_path=True,
-        tokenizer=tokenizer,
-    )
-    print(
-        f"Generated Onnx_path {generated_onnx_path} and Onnx_model_path {onnx_model_path} and Onnx_dir_path is {onnx_dir_path}"
-    )
-    assert (
-        generated_onnx_path == onnx_model_path
-    ), f"ONNX files were generated at an unusual location, expected {onnx_model_path}, got {generated_onnx_path}"
-    logger.info(f"Base Path is {base_path} and Onnx Model Path is : {generated_onnx_path}")
-
-    # Compile
-    # We need to pass parent directory of qpc_dir_path, as the compile function handles the qpcs directory creation
-    generated_qpc_path = compile(
-        onnx_path=onnx_model_path,
-        qpc_path=os.path.dirname(qpc_dir_path),
-        num_cores=num_cores,
-        batch_size=batch_size,
-        prompt_len=prompt_len,
-        ctx_len=ctx_len,
-        mxfp6=mxfp6,
-        mxint8=mxint8,
-        aic_enable_depth_first=aic_enable_depth_first,
-        mos=mos,
-        device_group=device_group,
-    )
-    assert (
-        qpc_dir_path == generated_qpc_path
-    ), f"QPC files were generated at an unusual location, expected {qpc_dir_path}; got {generated_qpc_path}"
-    logger.info(f"Compiled qpc files can be found at : {generated_qpc_path}")
+        assert (
+            qpc_dir_path == generated_qpc_path
+        ), f"QPC files were generated at an unusual location, expected {qpc_dir_path}; got {generated_qpc_path}"
+        logger.info(f"Compiled qpc files can be found at : {generated_qpc_path}")
+
 
     # Execute
     cloud_ai_100_exec_kv(
```

Comment:
Compilation and execution are called twice in this implementation. It would be better if we can remove the redundancy
Sample implementation
` if qpc_path_exists:
        # Continue to execute with pre-compiled QPC
        logger.info("Pre-compiled QPC found! Executing with given prompt.")
    else:
        # Check for ONNX model existence or transform and export HF model
        onnx_path_exists, onnx_dir_path, onnx_model_path = onnx_exists(model_name)
        
        # Check if onnx file exists
        if not onnx_path_exists:
            # Transform hf model -> export -> compile
            model_hf = AutoModelForCausalLM.from_pretrained(model_hf_path, use_cache=True)
            
            # Easy and minimal api to update the model to QEff.
            model_transformed = QEfficient.transform(model_hf, type="Transformers", form_factor="cloud")
            logger.info(f"Model after Optimized transformations {model_transformed}")

            # Export to ONNX
            logger.info(f"Exporting to Pytorch {model_name} to ONNX...")
            base_path, generated_onnx_path = qualcomm_efficient_converter(
                model_kv=model_transformed,
                onnx_dir_path=onnx_dir_path,
                model_name=model_name,
                kv=True,
                form_factor="cloud",
                return_path=True,
                tokenizer=tokenizer,
            )
            
            print(f"Generated Onnx_path {generated_onnx_path} and Onnx_model_path {onnx_model_path} and Onnx_dir_path is {onnx_dir_path}")

            assert (generated_onnx_path == onnx_model_path), f"ONNX files were generated at an unusual location, expected {onnx_model_path}, got {generated_onnx_path}"
            
            logger.info(f"Base Path is {base_path} and Onnx Model Path is : {generated_onnx_path}")
           
        # Compile the model and generate QPC
        # We need to pass parent directory of qpc_dir_path, as the compile function handles the qpc directory creation
        if not qpc_path_exists:
            generated_qpc_path = compile(
                onnx_path=onnx_model_path,
                qpc_path=os.path.dirname(qpc_dir_path),
                num_cores=num_cores,
                batch_size=batch_size,
                prompt_len=prompt_len,
                ctx_len=ctx_len,
                mxfp6=mxfp6,
                mxint8=mxint8,
                aic_enable_depth_first=aic_enable_depth_first,
                mos=mos,
                device_group=device_group,
            )

            assert (qpc_dir_path == generated_qpc_path), f"QPC files were generated at an unusual location, expected {qpc_dir_path}; got {generated_qpc_path}"
            logger.info(f"Compiled qpc files can be found at : {generated_qpc_path}")

    # Execute model on device
    cloud_ai_100_exec_kv(
        batch_size,
        tokenizer=tokenizer,
        qpc_path=qpc_dir_path,
        device_id=device_group,
        prompt=prompt,
    )`

Reply:
Please check addresssed this.